### PR TITLE
Add Publish-GcpsMessage to publish PubSub message to a topic.

### PR DESCRIPTION
**Pull request by quoctruong**
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-powershell/402)

### CODE REVIEW

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+        internal static PubsubMessage ConstructPubSubMessage(string data, Hashtable attributes)
+        {
+            Dictionary<string, string> attributesDict = null;
+            if (attributes != null && attributes.Count > 0)
+            {
+                attributesDict = ConvertToDictionary<string, string>(attributes);
+            }
+            // A valid PubSub message must have either a non-empty message or a non-empty attributes.
+            if (attributesDict == null && data == null)
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Should you also check for data being an empty string here?

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        /// The topic to which the messages will be published.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Topic { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// Data of the message that will be published.
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: "Data of the message.." this seems a little hard to parse.  Maybe "The data message.."?

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            {
+                PublishResponse response = publishRequest.Execute();
+
+                // The MessageIds is a server-assigned ID of the published message.
+                // They are in the same order as the messages in the request.
+                // We will use them to populate each message in the request with the corresponding message id.
+                IList<string> returnedMessageIds = response.MessageIds;
+                if (returnedMessageIds.Count != requestBody.Messages.Count)
+                {
+                    throw new InvalidOperationException($"Only {returnedMessageIds.Count} out of {requestBody.Messages.Count} is published.");
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: I think 'is' should be dropped.  

"Only 2 out of 10 published."   and "Only 1 out of 10 published."  

Otherwise I think you will need to use is for singular and are for multiple

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                }
+
+                for (int index = 0; index < returnedMessageIds.Count; index += 1)
+                {
+                    PubsubMessage publishedMessage = requestBody.Messages[index];
+                    publishedMessage.MessageId = returnedMessageIds[0];
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: returnedMessageIds[index];?

Or am I misunderstanding the message ids?  If so could you add a comment?

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                    if (!string.IsNullOrWhiteSpace(publishedMessage.Data))
+                    {
+                        byte[] decodedMessageDataBytes = Convert.FromBase64String(publishedMessage.Data);
+                        publishedMessage.Data = Encoding.UTF8.GetString(decodedMessageDataBytes);
+                    }
+                    WriteObject(publishedMessage);
+                    index += 1;
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Why the extra increment?

File: Google.PowerShell/PubSub/GcpsMessage.cs (hunk 1 above)
Comment (quoctruong) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Fixed.

File: Google.PowerShell/PubSub/GcpsMessage.cs (hunk 2 above)
Comment (quoctruong) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Fixed.

File: Google.PowerShell/PubSub/GcpsMessage.cs (hunk 3 above)
Comment (quoctruong) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Fixed.

File: Google.PowerShell/PubSub/GcpsMessage.cs (hunk 4 above)
Comment (quoctruong) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Fixed.

File: Google.PowerShell/PubSub/GcpsMessage.cs (hunk 5 above)
Comment (quoctruong) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Removed.

File: Google.PowerShell/PubSub/GcpsMessage.cs [hunk 6]
```diff
@@ -194,20 +194,19 @@ protected override void ProcessRecord()
...
                 if (returnedMessageIds.Count != requestBody.Messages.Count)
                 {
-                    throw new InvalidOperationException($"Only {returnedMessageIds.Count} out of {requestBody.Messages.Count} is published.");
+                    throw new InvalidOperationException($"Only {returnedMessageIds.Count} out of {requestBody.Messages.Count} published.");
                 }
 
                 for (int index = 0; index < returnedMessageIds.Count; index += 1)
                 {
                     PubsubMessage publishedMessage = requestBody.Messages[index];
-                    publishedMessage.MessageId = returnedMessageIds[0];
```
Comment (iantalarico) [inline Google.PowerShell/PubSub/GcpsMessage.cs]: Should there be some more intrusive tests here?  I'm surprised neither of these  caused issues.

**Comment by quoctruong** [comment]
Thanks!
